contracts/amm/SirenExchange.sol

```diff
@@ -0,0 +1,280 @@
+        address[] calldata path,
+        uint256 tokenAmountInMaximum,
+        address sirenAmmAddress,
+        uint256 deadline,
+        address _router
+    ) external returns (uint256[] memory amounts) {
+        uint256 collateralPremium = ISirenTradeAMM(sirenAmmAddress)
+            .bTokenGetCollateralIn(seriesId, bTokenAmount);
+
+        uint256[] memory amountsIn = IUniswapV2Router02(_router).getAmountsIn(
+            collateralPremium,
+            path
+        );
+
+        require(amountsIn[0] <= tokenAmountInMaximum, "Not Enough tokens sent");
+
+        TransferHelper.safeTransferFrom(
+            path[0],
+            msg.sender,
+            address(this),
+            amountsIn[0]
+        );
+        TransferHelper.safeApprove(
+            path[0],
+            address(IUniswapV2Router02(_router)),
+            amountsIn[0]
+        );
+
+        // Executes the swap, returning the amountIn actually spent.
+        amounts = IUniswapV2Router02(_router).swapTokensForExactTokens(
+            collateralPremium,
+            amountsIn[0],
+            path,
+            address(this),
+            deadline
+        );
+
+        TransferHelper.safeApprove(
+            path[path.length - 1],
+            sirenAmmAddress,
+            collateralPremium
+        );
+
+        //Call MinterAmm bTokenBuy contract
+        ISirenTradeAMM(sirenAmmAddress).bTokenBuy(
+            seriesId,
+            bTokenAmount,
+            collateralPremium
+        );
+
+        //Transfer the btokens to the correct address ( caller of this contract)
+        erc1155Controller.safeTransferFrom(
+            address(this),
+            msg.sender,
+            SeriesLibrary.bTokenIndex(seriesId),
+            bTokenAmount,
+            dataReturn()
+        );
+
+        emit BTokenBuy(
+            amounts,
+            path,
+            sirenAmmAddress,
+            bTokenAmount,
+            seriesId,
+            msg.sender
+        );
+
+        return amounts;
+    }
+
+    /// @notice Sell the wToken of a given series to the AMM in exchange for collateral token
+    /// @param seriesId The ID of the Series to buy wToken on
+    /// @param bTokenAmount The amount of bToken to sell (bToken has the same decimals as the underlying)
+    /// @param path The path of the collateral token of the series to the collateral the caller wishes to receive
+    /// @param tokenAmountOutMinimum The lowest amount of collateral the caller is willing to receive as payment
+    /// @param sirenAmmAddress address of the amm that we wish to call
+    /// @param deadline deadline the transaction must be completed by
+    /// @param _router address of the router we wish to use ( QuickSwap or SushiSwap )
+    /// We supply a bToken and then select which collateral we wish to receive as our payment ( if it isnt the underlying asset )
+    function bTokenSell(
+        uint64 seriesId,
+        uint256 bTokenAmount,
+        address[] calldata path,
```

Comment:
Are there any sanity checks that need to be done to the paths before they are used, or are we assuming the exchange will fail if it is not valid?

Reply:
I believe we can assume the trade would fail if its not valid but let me look at this again and make sure it wont go through. If you provide a correct path ( one that exists on the router we are. using ) it could get through the "exchange" piece of this however then it would need to get through the sell or buy section... which if it did would we care.

Worst case would be a sender sent in tokens that had a valid path on the router and then receive btokens from that.... I'll give this some more thought.

Let me know if this comment makes sense.